src/finch/interface/lazy.py

```diff
@@ -932,6 +952,197 @@ def vecdot(x1, x2, /, *, axis=-1) -> LazyTensor:
+        raise ValueError(
+            f"Tensor with shape {tensor.shape} is not broadcastable "
+            f"to the shape {shape}"
+        )
+    return elementwise(first, tensor, NoneArray(shape))
+
+
+def broadcast_arrays(*arrays: LazyTensor) -> tuple[LazyTensor, ...]:
+    """
+    Broadcasts one or more arrays against one another.
+    """
+    shape = _broadcast_shape(*(array.shape for array in arrays))
+    return tuple(broadcast_to(arr, shape) for arr in arrays)
+
+
+def concat(arrays: tuple | list, /, axis: int | None = 0) -> LazyTensor:
+    """
+    Concatenates input tensors along the specified axis using identity matrix placement.
+
+    Parameters:
+        arrays: Sequence of tensors to concatenate (tuple or list of LazyTensor)
+        axis: Axis along which to concatenate (default=0)
+
+    Returns:
+        Concatenated tensor as LazyTensor
+    """
+    arrays = [defer(arr) for arr in arrays]
+
+    if axis is None:
+        # flatten all tensors
+        for i, t in enumerate(arrays):
+            arrays[i] = flatten(t)
+        axis = 0
+
+    # Convert axis to positive index and validate
+    ndim = arrays[0].ndim
+    axis = normalize_axis_index(axis, ndim)
+
+    # Validate shapes and compute total size
+    sizes = []  # used to build shift matrices
+    total_size = 0
+
+    for t in arrays:
+        if t.ndim != ndim:
+            raise ValueError("All tensors must have same number of dimensions")
+        sizes.append(t.shape[axis])
+        total_size += t.shape[axis]
+
+    # Helper to create numpy shift matrix
+    def build_shift_matrix(offset: int, size: int) -> LazyTensor:
+        """Creates a lazy shift matrix that positions tensor at given offset"""
+        mat = np.zeros((size, total_size), dtype=np.int8)
+        for j in range(size):
+            mat[j, offset + j] = 1
+        return defer(mat)
+
+    # Build result by summing positioned tensors
+    cumulative_offset = 0
+    result = None
+
+    for i, tensor in enumerate(arrays):
+        size = sizes[i]
+        I_i = build_shift_matrix(cumulative_offset, size)
+
+        # Do a tensordot but with different operations
+        # permute tensor
+        axis_permutation = tuple(i for i in range(tensor.ndim) if i != axis) + (axis,)
+        permuted_tensor = permute_dims(tensor, axis_permutation)
+        # add 1-dim at the end of permuted_tensor for broadcasting
+        permuted_tensor = expand_dims(permuted_tensor, -1)
+
+        # Contract tensor with identity matrix along the concatenation axis
+        placed = reduce(
+            logical_or,
+            elementwise(choose_if, permuted_tensor, I_i),
+            axis=-2,
+        )
+        # permute back
+        placed_perm = moveaxis(placed, -1, axis)
+
+        # Initialize or accumulate result
+        result = placed_perm if result is None else result + placed_perm
+
+        cumulative_offset += size
+    assert result is not None
+    return result
+
+
+def flatten(x):
```

Comment:
Similar to Concat, could you define `SplitDimsTensor` and `CombineDimsTensor`? Then you can define flatten as a `CombineDimsTensor`, and we have a good shot of supporting reshape.

Reply:
you don't need to support reshape now, i'd like to handle that myself.